src/bmp/BMPParser.cc

```diff
@@ -157,39 +156,42 @@ void Parser::parse(uint8_t *buf, int bufSize, uint8_t *format){
     skip(8);
 
     // Number of colors in the palette or 0 if no palette is present
-    auto palColNum = U4();
-    EU(palColNum, "non-empty color palette");
+    palColNum = U4();
 
-    // Number of important colors used or 0 if all colors are important
-    auto impCols = U4();
-    EU(impCols, "non-zero important colors");
+    // Number of important colors used or 0 if all colors are important (generally ignored)
+    skip(4);
 
-    // BITMAPV4HEADER has additional properties
-    if(dibSize == 108){
+    if(infoHeader >= 2){
       // If BI_BITFIELDS are used, calculate masks, otherwise ignore them
       if(compr == 3){
         // Convert each mask to bit offset for faster shifting
-        CALC_MASK(red);
-        CALC_MASK(green);
-        CALC_MASK(blue);
-        CALC_MASK(alpha);
+        calcMaskShift(redShift, redMask, redMultp);
+        calcMaskShift(greenShift, greenMask, greenMultp);
+        calcMaskShift(blueShift, blueMask, blueMultp);
+        calcMaskShift(alphaShift, alphaMask, alphaMultp);
       }else{
         skip(16);
       }
 
-      // Encure that the color space is LCS_WINDOWS_COLOR_SPACE
-      string colSpace = getStr(4, 1);
-      EU(colSpace != "Win ", "color space \"" + colSpace + "\"");
-
-      // The rest 48 bytes are ignored for LCS_WINDOWS_COLOR_SPACE
-      skip(48);
+      if(infoHeader >= 4){
+        if(!palColNum){
+          // Ensure that the color space is LCS_WINDOWS_COLOR_SPACE or sRGB
+          string colSpace = getStr(4, 1);
+          EU(colSpace != "Win " && colSpace != "sRGB", "color space \"" + colSpace + "\"");
+        }else{
+          skip(4);
+        }
+
+        // The rest 48 bytes are ignored for LCS_WINDOWS_COLOR_SPACE and sRGB
+        skip(48);

```

Comment:
This also needs to be conditioned on the header size, right? (it could be 108 or 124 bytes at this point)

Reply:
I'm not sure exactly what you mean. The only difference between V4 and V5 is that V5 has color management profiles, but they are located after the palette (and generally ignored), we already skip them in the line `ptr = data + imgdOffset`, so neither color palette nor image data would be the problem. Or am I missing something?

Reply:
I could be mistaken, but what I meant here and on the prior comment is that I thought the V2 header is 4B smaller than the V3 header, so there should be no `skip(4)` for V2 after parsing the mask for blue; and that V4 is 16B smaller than V5, so this skip is 32 for V4. Am I confused? You know the format better than I do now.

Reply:
Yes, that would affect only V1 nd V2 images with color palette (bit I couldn't find an example of such image). Since we know that palette immediately follows DIB header, I simplified the code (now `skip(4)` wouldn't affect parsing, but I removed it anyway).

> so this skip is 32 for V4

It's not relevant anymore, but I think we would want to skip 48 for V4 and 64 for V5 (tested empirically on some images).

Reply:
There's a suite of BMPs here that includes one (q/rgb32h52.bmp) with a V2INFOHEADER: http://entropymine.com/jason/bmpsuite/ and http://entropymine.com/jason/bmpsuite/bmpsuite/html/bmpsuite.html and your parser works on that file, congrats 🙂! Firefox doesn't. Most images seem to work with just a few exceptions (g/rgb32.bmp for example, and a few I created with "advance settings" in Photoshop that cause "inconsistent image data size" messages).

---

I think I get what you're saying now: the `skip()` doesn't matter because you seek to an absolute position with the `ptr = data + imgdOffset` line right after. Previously it might have skipped into the palette but that didn't actually affect parsing. 👍